vicinity/backends/faiss.py

```diff
@@ -0,0 +1,229 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Literal
+
+import faiss
+import numpy as np
+from numpy import typing as npt
+
+from vicinity.backends.base import AbstractBackend, BaseArgs
+from vicinity.datatypes import Backend, QueryResult
+from vicinity.utils import normalize
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class FaissArgs(BaseArgs):
+    dim: int = 0
+    index_type: Literal["flat", "ivf", "hnsw", "lsh", "scalar", "pq", "ivf_scalar", "ivfpq", "ivfpqr"] = "hnsw"
+    metric: Literal["cosine", "l2"] = "cosine"
+    nlist: int = 100
+    m: int = 8
+    nbits: int = 8
+    refine_nbits: int = 8
+    direct_map: bool = True
+
+
+class FaissBackend(AbstractBackend[FaissArgs]):
+    argument_class = FaissArgs
+
+    def __init__(
+        self,
+        index: faiss.Index,
+        arguments: FaissArgs,
+    ) -> None:
+        """Initialize the backend using a FAISS index."""
+        super().__init__(arguments)
+        self.index = index
+        # Enable DirectMap if specified and supported by index type
+        if isinstance(index, faiss.IndexIVF) and arguments.direct_map:
```

Comment:
so setting direct_map to `True` doesn't do anything for any index besides `IVF`. Maybe flag it or crash? 

Do you ever want to set it to False? If not, you could just not have it as an argument and enable it for IVF.

Reply:
Good point, let's just default it to true.